pkg/service/storage/handlers/replica/transfer.go

```diff
@@ -223,13 +391,74 @@ func Transfer(ctx context.Context, service TransferService, request *TransferReq
 	if err != nil {
 		return fmt.Errorf("failed to send request for receipt: %w", err)
 	}
+
 	if uploadServiceResponse.Status() >= 300 || uploadServiceResponse.Status() < 200 {
 		topErr := fmt.Errorf("unsuccessful http POST to upload service")
 		resData, err := io.ReadAll(uploadServiceResponse.Body())
 		if err != nil {
-			return fmt.Errorf("%s failed to read replication sink response body: %w", topErr, err)
+			return fmt.Errorf("%s failed to read response body: %w", topErr, err)
 		}
-		return fmt.Errorf("%s response body: %s: %w", topErr, resData, err)
+		return fmt.Errorf("%s response body: %s", topErr, resData)
+	}
+
+	return nil
+}
+
+// TransferReplicaBlobError represents a transfer operation failure that can be encoded to IPLD
+type TransferReplicaBlobError struct {
+	message string
+}
+
+func (tf TransferReplicaBlobError) Name() string {
+	return "TransferReplicaBlobError"
+}
+
+func (tf TransferReplicaBlobError) Error() string {
+	return tf.message
+}
+
+func (tf TransferReplicaBlobError) ToIPLD() (ipld.Node, error) {
+	name := tf.Name()
+	model := struct {
+		Name    *string `json:"name"`
+		Message string  `json:"message"`
+	}{
+		Name:    &name,
+		Message: tf.message,
+	}
+	return ipld.WrapWithRecovery(&model, nil)
+}
+
+// SendFailureReceipt sends a failure receipt to the upload service when Transfer fails after all retries
+func SendFailureReceipt(ctx context.Context, service TransferService, request *TransferRequest, transferErr error) error {
+	// Create a failure that implements IPLDBuilder
+	failure := TransferReplicaBlobError{
+		message: fmt.Sprintf("Transfer failed after all retries: %v", transferErr),
+	}
+
+	// Create an error result for the transfer
+	errResult := result.Error[replica.TransferOk, TransferReplicaBlobError](failure)
+
+	// Issue the failure receipt
+	rcpt, err := receipt.Issue(service.ID(), errResult, ran.FromInvocation(request.Cause))
+	if err != nil {
+		return fmt.Errorf("issuing failure receipt: %w", err)
+	}
+
+	// Store the failure receipt
+	if err := service.Receipts().Put(ctx, rcpt); err != nil {
+		return fmt.Errorf("failed to store failure receipt: %w", err)
+	}
+
+	// Create agent message with the failure receipt
+	agentMsg, err := message.Build([]invocation.Invocation{}, []receipt.AnyReceipt{rcpt})
+	if err != nil {
+		return fmt.Errorf("building failure message: %w", err)
+	}
+
+	// Send the failure receipt to the upload service
+	if err := sendMessageToUploadService(ctx, service, agentMsg); err != nil {
+		return fmt.Errorf("sending failure receipt: %w", err)
```

Comment:
@alanshaw I am uncertain of this `TransferReplicaBlobError` type implementation. I _think_ it might require a change in go-libstoracha, but I am not sure. wdyt?